### PR TITLE
Protocol 5

### PR DESCRIPTION
**Pull request by parv0888**
## Purpose
Added support for Contract Upgraded Event (Protocol 5)

## Changes
Added support for Contract Upgraded Event (Protocol 5)

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

## CLA acceptance

By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [x] I accept the above linked CLA.

### CODE REVIEW

**Comment by parv0888** [comment]
image

File: backend/Tests/Api/GraphQL/Import/TransactionsWriterTest.cs [hunk 1]
```diff
@@ -612,6 +612,29 @@ public async Task TransactionEvents_ContractResumed()
...
+    [Fact]
+    public async Task TransactionEvents_ContractUpgraded()
+    {
+        _blockSummaryBuilder
+            .WithTransactionSummaries(new TransactionSummaryBuilder()
+                .WithResult(new TransactionSuccessResultBuilder()
+                    .WithEvents(new Upgraded(
+                        new ContractAddress(1423, 1),
+                        new ModuleRef("73ba390d9ce2bb1bf54f124bb00e9dee0d6dc40d6de0f5ba06e1d1f095e4afcc"),
+                        new ModuleRef("73ba390d9ce2bb1bf54f124bb00e9dee0d6dc40d6de0f5ba06e1d1f095e4afcc")
```
Comment (abizjak) [inline backend/Tests/Api/GraphQL/Import/TransactionsWriterTest.cs:625]: I recommend using a different hash for from and to in the test. Just 
```
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
for example.

File: backend/Tests/Api/GraphQL/Import/TransactionsWriterTest.cs [hunk 2]
```diff
@@ -612,6 +612,29 @@ public async Task TransactionEvents_ContractResumed()
...
+                    ))
+                    .Build())
+                .Build());
+        
+        await WriteData();
+
+        var result = await ReadSingleTransactionEventType<ContractUpgraded>();
+        result.ContractAddress.Should().Be(new Application.Api.GraphQL.ContractAddress(1423, 1));
+        result.From.Should().Be("73ba390d9ce2bb1bf54f124bb00e9dee0d6dc40d6de0f5ba06e1d1f095e4afcc");
+        result.To.Should().Be("73ba390d9ce2bb1bf54f124bb00e9dee0d6dc40d6de0f5ba06e1d1f095e4afcc");
```
Comment (abizjak) [inline backend/Tests/Api/GraphQL/Import/TransactionsWriterTest.cs:635]:
```suggestion
        result.To.Should().Be("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");
```

File: backend/Tests/Api/GraphQL/Import/TransactionsWriterTest.cs (hunk 1 above)
Comment (abizjak) [inline backend/Tests/Api/GraphQL/Import/TransactionsWriterTest.cs:625]:
```suggestion
                        new ModuleRef("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
```

File: backend/Tests/ConcordiumSdk/NodeApi/JsonConverters/TransactionResultEventConverterTest.cs [hunk 3]
```diff
@@ -400,6 +400,26 @@ public void RoundTrip_Resumed()
...
     }
 
+    [Fact]
+    public void RoundTrip_Upgraded()
+    {
+        var json = @"{
+						""address"": { ""index"": 1039, ""subindex"": 0 },
+						""from"": ""73ba390d9ce2bb1bf54f124bb00e9dee0d6dc40d6de0f5ba06e1d1f095e4afcc"",
+						""tag"": ""Upgraded"",
+						""to"": ""73ba390d9ce2bb1bf54f124bb00e9dee0d6dc40d6de0f5ba06e1d1f095e4afcc""
```
Comment (abizjak) [inline backend/Tests/ConcordiumSdk/NodeApi/JsonConverters/TransactionResultEventConverterTest.cs:410]:
```suggestion
						""to"": ""aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa""
```

File: backend/Tests/ConcordiumSdk/NodeApi/JsonConverters/TransactionResultEventConverterTest.cs [hunk 4]
```diff
@@ -400,6 +400,26 @@ public void RoundTrip_Resumed()
...
+					}";
+        
+        var deserialized = JsonSerializer.Deserialize<TransactionResultEvent>(json, _serializerOptions);
+        var typed = Assert.IsType<Upgraded>(deserialized);
+        typed.Address.Should().Be(new ContractAddress(1039, 0));
+        typed.From.Should().Be(new ModuleRef("73ba390d9ce2bb1bf54f124bb00e9dee0d6dc40d6de0f5ba06e1d1f095e4afcc"));
+        typed.To.Should().Be(new ModuleRef("73ba390d9ce2bb1bf54f124bb00e9dee0d6dc40d6de0f5ba06e1d1f095e4afcc"));
```
Comment (abizjak) [inline backend/Tests/ConcordiumSdk/NodeApi/JsonConverters/TransactionResultEventConverterTest.cs:417]:
```suggestion
        typed.To.Should().Be(new ModuleRef("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"));
```

**Review by abizjak — APPROVED** [review]
Please merge when resolving the comments about tests.

File: backend/Tests/Api/GraphQL/Import/TransactionsWriterTest.cs (hunk 1 above)
Comment (parv0888) [reply backend/Tests/Api/GraphQL/Import/TransactionsWriterTest.cs]: done

Comment (parv0888) [reply backend/Tests/Api/GraphQL/Import/TransactionsWriterTest.cs]: done

File: backend/Tests/ConcordiumSdk/NodeApi/JsonConverters/TransactionResultEventConverterTest.cs (hunk 4 above)
Comment (parv0888) [reply backend/Tests/ConcordiumSdk/NodeApi/JsonConverters/TransactionResultEventConverterTest.cs]: done

**Pull request merged**